# Analyzing Used Cars data on eBay (Kleinanzeigen)

We will be working on a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io/) including simulating a less-cleaned version of the data.


The aim of this project is to clean the data and analyze the included used car listings.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [315]:
import pandas as pd
import numpy as np

In [316]:
autos = pd.read_csv("./resources/autos.csv", encoding="Latin-1")

autos.info()
autos.head()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50


The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

* `name` - Name of the car.

* `seller` - Whether the seller is private or a dealer.

* `offerType` - The type of listing

* `price` - The price on the ad to sell the car.

* `abtest` - Whether the listing is included in an A/B test.

* `vehicleType` - The vehicle Type.

* `yearOfRegistration` - The year in which the car was first registered.

* `gearbox` - The transmission type.

* `powerPS` - The power of the car in PS.

* `model` - The car model name.

* `odometer` - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.

* `fuelType` - What type of fuel the car uses.

* `brand` - The brand of the car.

* `notRepairedDamage` - If the car has a damage which is not yet repaired.

* `dateCreated` - The date on which the eBay listing was created.

* `nrOfPictures` - The number of pictures in the ad.

* `postalCode` - The postal code for the location of the vehicle.

* `lastSeenOnline` - When the crawler saw this ad last online.


Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

## Clean column names


In [317]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

* Change the columns from camelcase to snakecase.
* Change a few wordings to more accurately describe the columns.

In [318]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "abtest", "vehicle_type", "registration_year", "gearbox", "power_ps", "model", "odometer", "registration_month", "fuel_type", "brand", "unrepaired_damage", "ad_created", "nr_pictures", "postal_code", "last_seen"]


autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

* Any columns that need more investigation

* Any examples of numeric data stored as text that needs to be cleaned




In [319]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

We can see that:

* `Registration year`, have a min and max year that don't look correct, it needs investigation;

* `number of pictures` are always zero, this column can be removed;

* `Odometer` and `price` are stored as string, we will convert it to number;


Let's start of with `Price` and `Odometer` columns.
We're going to:

* Remove any non-numeric characters.
* Convert the column to a numeric dtype.
* Use DataFrame.rename() to rename the column to odometer_km


In [320]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)
#autos["price"].head()
#autos["price"].value_counts()

autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",","").astype(int)
#autos["odometer"].head()

/var/folders/w4/23b90vln373g35yg9v6lmwv80000gn/T/ipykernel_99639/2098876550.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)


In [321]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

autos.head()




date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Regarding the values on the columns, there are a number of text columns where all (or nearly all) of the values are the same:

* `seller`
* `offer_type`
* `nr_pictures`

In [322]:
print("seller: \n {}".format(autos["seller"].value_counts()))

print("Offer_type: \n {}".format(autos["offer_type"].value_counts()))

print("Pictures: \n {}".format(autos["nr_pictures"].value_counts()))

seller: 
 privat        49999
gewerblich        1
Name: seller, dtype: int64
Offer_type: 
 Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
Pictures: 
 0    50000
Name: nr_pictures, dtype: int64


As we can see above, all the values are the same on these 3 columns. Therefore, we can remove them from our dataset, since they don't add any value to our analysis

In [323]:
autos = autos.drop(["seller", "offer_type", "nr_pictures"], axis=1)
# autos.head() to ensure it's removed

## Exploring the Odometer and Price columns

Now, let's analyse the values these two columns.

We'll be analyzing the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

We'll be using the following methods:

* Series.unique().shape to see how many unique values

* Series.describe() to view min/max/median/mean etc

* Series.value_counts(), with some variations:

	* chained to .head() if there are lots of values.

	* Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).
	
* When removing outliers, we can do df[(df["col"] >= x ) & (df["col"] <= y )], but it's more readable to use df[df["col"].between(x,y)]






In [324]:
print("Unique values: {}".format(autos["odometer_km"].unique().shape))
autos["odometer_km"].value_counts()


Unique values: (13,)


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded and there are only 13 values, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [325]:
print("Price unique values: {}".format(autos["price"].unique().shape))


Price unique values: (2357,)


However, there are 2357 unique entries of prices, which might indicate the user type it's individual car value. Let's dig deeper and analyse the maximum and minimum values of the prices

In [326]:
print("{}".format(autos["price"].describe()))

print("Lowest prices: \n{}".format(autos["price"].value_counts().head(15)))

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
Lowest prices: 
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64


We can see that we have 1421 cars without price (zero), which we'll remove.

Let's take a look at the highest price outbound now:

In [327]:
print("Highest prices: \n{}".format(autos["price"].value_counts().sort_index(ascending=False).head(20)))

Highest prices: 
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


In [328]:
print("Lowest prices: \n{}".format(autos["price"].value_counts().sort_index(ascending=True).head(20)))


Lowest prices: 
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


There are a number of listings with prices below `30`. There are also a small number of listings with very high values, including 14 at around or over `$1 million`.

Given that eBay is an auction site, there could legitimately be items where the opening bid is 
1 . We will keep these items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers, especially for used cars.

In [329]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website


Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [330]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use Series.str[:10]:

In [331]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


In [332]:
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [333]:
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values())

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.



In [334]:
(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [335]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [336]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.



## Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [337]:
autos["registration_year"].sort_values(ascending=True).head(20)

22316    1000
49283    1001
24511    1111
32585    1800
10556    1800
22659    1910
45157    1910
30781    1910
28693    1910
3679     1910
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
39725    1937
21421    1937
23804    1937
26607    1937
26103    1938
Name: registration_year, dtype: int64

Since the first car invented and patented was in 1886, by Carl Benz, we will ignore every date lower than that.

Thus, our lowest bound date is going to be 1910.

Let's determine what percentage of our data has invalid values in this column, below `1900` and above `2016`



In [338]:
wrong_years = (~autos["registration_year"].between(1900,2016)).sum() 

print("There are {} entries with years above 2016 or below 1900".format(wrong_years))

print("That corresponds to {:.2f}% percentage of our sample".format((wrong_years / autos.shape[0])*100))

There are 1884 entries with years above 2016 or below 1900
That corresponds to 3.88% percentage of our sample


Since it corresponds to less than 4% of our sample, we will remove this data from our dataset

In [339]:
autos = autos[autos["registration_year"].between(1900, 2016)]

autos["registration_year"].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1939    0.000021
1948    0.000021
1938    0.000021
1953    0.000021
1943    0.000021
Name: registration_year, Length: 78, dtype: float64

The number of years has been reduced to 78 entries, with the most expression taking part in the early 2000's.

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the `brand` column.

Let's explore the unique values in the brand column, and decide on which brands we want to aggregate by.

In [340]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [341]:
top_20_brands_expression = autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(17).sum()

print("The top 17 brands advertised on Ebay totalize a {:.2f}% of the market".format(top_20_brands_expression * 100))

The top 17 brands advertised on Ebay totalize a 90.05% of the market


Assuming that the most 17 brands publicized on ebay account for ~90% of the market, we will simply consider these to our analysis.

In [342]:
top_brands = autos.loc[autos["brand"].isin(autos["brand"].value_counts(normalize=True)[autos["brand"].value_counts(normalize=True) >= 0.01].index), "brand"].unique()

print(top_brands)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'seat' 'renault'
 'mercedes_benz' 'audi' 'opel' 'mazda' 'toyota' 'nissan' 'fiat' 'skoda'
 'citroen' 'hyundai']


With the above list of cars, let's explore the average price praticed per brand.

In [343]:
avg_brand_price = {}

for b in top_brands:
	b_only = autos[autos["brand"] == b]
	mean_price = b_only["price"].mean()
	avg_brand_price[b] = int(mean_price)

sorted_price_brand = dict(sorted(avg_brand_price.items(), key=lambda x: x[1], reverse=True))
print(sorted_price_brand)
print('\n')

autos["brand"].value_counts().head(10)

{'audi': 9336, 'mercedes_benz': 8628, 'bmw': 8332, 'skoda': 6368, 'volkswagen': 5402, 'hyundai': 5365, 'toyota': 5167, 'nissan': 4743, 'seat': 4397, 'mazda': 4112, 'citroen': 3779, 'ford': 3749, 'smart': 3580, 'peugeot': 3094, 'opel': 2975, 'fiat': 2813, 'renault': 2474}




volkswagen       9862
bmw              5137
opel             5022
mercedes_benz    4503
audi             4041
ford             3263
renault          2201
peugeot          1393
fiat             1197
seat              853
Name: brand, dtype: int64

* We can see that Audi has around 4k cars with an average price of $9.336, which makes it the most expensive brand by average. In second comes Mercedes, immediately followed by BMW. 

* Renault, Fiat, Opel and Ford come as the cheapest brands to acquire. 

* The most advertised brand in Ebay, Volkswage has an average price of around $5.5K, a great solution for someone with a more conservative bugdet yet still requiring a reliable car

## Exploring Mileage

Since there are few cars to correlate the price with the mileage, we will just use the top commercialised brands.
Thus, for the top 6 brands, we'll use aggregation to understand the average mileage for those cars and if there's any visible link with `mean`/ average price.

We will create a new Dataframe with both values combined.

* Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.

* Convert both dictionaries to series objects, using the series constructor.

- Create a dataframe from the first series object using the dataframe constructor.

- Assign the other series as a new column in this dataframe.

- Pretty print the dataframe, and write a paragraph analyzing the aggregate data.

In [344]:
top_brand_mean_mileage = {}

top_brand_price_mean = {}

top_6_brands = list(autos["brand"].value_counts().head(6).index)
print("Top 6 brands: \n{}\n".format(top_6_brands))

for brand in top_6_brands:
	brand_only = autos[autos["brand"] == brand]
	mean_mileage = int(brand_only["odometer_km"].mean())

	top_brand_mean_mileage[brand] = mean_mileage
	top_brand_price_mean[brand] = avg_brand_price[brand]


mileage_series = pd.Series(top_brand_mean_mileage).sort_values(ascending=False)
price_series = pd.Series(top_brand_price_mean)

brand_top_df = pd.DataFrame(mileage_series, columns=["mileage_mean"])
brand_top_df["average_price"] = price_series

brand_top_df

Top 6 brands: 
['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']



mileage_mean  average_price
bmw                  132572           8332
mercedes_benz        130788           8628
opel                 129310           2975
audi                 129157           9336
volkswagen           128707           5402
ford                 124266           3749

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

In conclusion, if the user is looking for an cheap option amongst the most offered on ebay, they should search for Opel or Ford brands.
If on the other hand, they're looking for a more premium car and higher budget, the BMW, Benz or Audi are the most offered options